app/gui2/src/components/lexical/formatting.ts

```diff
@@ -56,3 +83,149 @@ function useFormatProperty(
+  editor: LexicalEditor,
+  onReadSelection: ($readSelection: (selection: RangeSelection) => void) => void,
+) {
+  const state = ref<BlockType>('paragraph')
+
+  onReadSelection((selection) => (state.value = $getBlockType(selection) ?? 'paragraph'))
+
+  const smallestEnabledHeading = ['h6', 'h5', 'h4', 'h3', 'h2', 'h1'].find(
+    (type) => type in blockTypeToBlockName,
+  ) as keyof typeof blockTypeToBlockName
+
+  function $getBlockType(selection: RangeSelection): BlockType | undefined {
+    const anchorNode = selection.anchor.getNode()
+    const element =
+      anchorNode.getKey() === 'root' ?
+        anchorNode
+      : $findMatchingParent(anchorNode, (e) => {
+          const parent = e.getParent()
+          return parent !== null && $isRootOrShadowRoot(parent)
+        }) ?? anchorNode.getTopLevelElementOrThrow()
+
+    if ($isListNode(element)) {
+      const parentList = $getNearestNodeOfType<ListNode>(anchorNode, ListNode)
+      const type = parentList ? parentList.getListType() : element.getListType()
+      if (type in blockTypeToBlockName) {
+        return type as keyof typeof blockTypeToBlockName
+      }
+    } else if ($isHeadingNode(element)) {
+      const type = element.getTag()
+      if (type in blockTypeToBlockName) {
+        return type as keyof typeof blockTypeToBlockName
+      } else {
+        return smallestEnabledHeading
+      }
+    } else {
+      const type = element.getType()
+      if (type in blockTypeToBlockName) {
+        return type as keyof typeof blockTypeToBlockName
+      }
+    }
+  }
+
+  const $setBlockType: Record<BlockType, () => void> = {
+    bullet: () => editor.dispatchCommand(INSERT_UNORDERED_LIST_COMMAND, undefined),
+    number: () => editor.dispatchCommand(INSERT_ORDERED_LIST_COMMAND, undefined),
+    paragraph: () => {
+      const selection = $getSelection()
+      if ($isRangeSelection(selection)) {
+        $setBlocksType(selection, () => $createParagraphNode())
+      }
+    },
+    quote: () => $setBlocksType($getSelection(), () => $createQuoteNode()),
+    code: () => {
+      let selection = $getSelection()
+      if (selection !== null) {
+        if (selection.isCollapsed()) {
+          $setBlocksType(selection, () => $createCodeNode())
+        } else {
+          const textContent = selection.getTextContent()
+          const codeNode = $createCodeNode()
+          selection.insertNodes([codeNode])
+          selection = $getSelection()
+          if ($isRangeSelection(selection)) {
+            selection.insertRawText(textContent)
+          }
+        }
+      }
+    },
+    h1: () => $setBlocksType($getSelection(), () => $createHeadingNode('h1')),
+    h2: () => $setBlocksType($getSelection(), () => $createHeadingNode('h2')),
+    h3: () => $setBlocksType($getSelection(), () => $createHeadingNode('h3')),
+  }
+
+  return computed({
+    get: () => state.value,
+    set: (value) => {
+      if (value !== state.value) editor.update($setBlockType[value])
+    },
+  })
```

Comment:
This is a weird pattern, having side effects in the computed setter. Why not return the `state` here, and create a watcher to run the update?

Reply:
That would lose the direction of data flow; the command should only be sent to the editor if the change didn't come from the editor. It's a good idea though to add buffering so that getters and setters are consistent, and to avoid spurious updates. Implemented `useBufferedWritable`.